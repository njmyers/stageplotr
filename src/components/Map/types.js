import type { Kit } from '../Kit/types';

export type Map = {
  name: string,
  width: number,
  height: number,
  length: number,
  items: Array<Kit>,
};
