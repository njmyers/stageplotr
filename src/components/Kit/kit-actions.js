// @flow

type Keys =
  | 'positionX'
  | 'positionY'
  | 'scale'
  | 'rotation'
  | 'layer'
  | 'orientation';

export type Action = {
  type: '@KIT/UPDATE_PROPERTY',
  key: Keys,
  value: number,
};

type Validator = (value: number) => boolean;

/**
 * Use this to create kit action creators
 * Sounds silly but it's just specifying which prop to update
 * Also you can pass in a validator function to give us warnings in development
 * Do we want stricter warnings?
 */
const kitActionCreator = (validator: Validator) => (key: Keys) => (
  value: number
): Action => {
  if (process.env.NODE_ENV !== 'production' && !validator((value: number))) {
    console.warn(`You are trying to set a kit property with an invalid value.
      key: ${key}
      value: ${value.toString()}
      validator: ${validator.name}`);
  }

  return {
    type: '@KIT/UPDATE_PROPERTY',
    key,
    value,
  };
};

/**
 * Validates position
 * We may need more validation here at the moment lets' keep it above 0
 */
const isValidPositionX: Validator = (value) => value > 0;

/**
 * Validates position
 * We may need more validation here at the moment lets' keep it above 0
 */
const isValidPositionY: Validator = (value) => value > 0;

/**
 * Validates the scale parameter
 * Scale should be between this range
 * This is a whole number percentage value
 * Will end up like transform: scale(value / 100)
 */
const isValidScale: Validator = (value) => value > 20 && value < 200;

/**
 * Validates a degree as a value between 0 and 360;
 * We can manipulate these in css using transform: rotate(deg)
 */
const isValidRotation: Validator = (value) => value >= 0 && value <= 360;

/**
 * Validates layer value
 * We can add in later if we need a minimum zIndex to keep above the map
 * This just checks it is below maximum zIndex value
 * https://stackoverflow.com/questions/491052/minimum-and-maximum-value-of-z-index
 */
const isValidLayer: Validator = (value) => value >= 0 && value <= 16777271;

/**
 * This validates a correct orientation value
 * Orientation is either 0 or -1.
 * Seems weird but the best way to do css flip is with transform: rotateX(-1)
 * So we can simply add these values with our scale values
 * https://css-tricks.com/snippets/css/flip-an-image/
 */
const isValidOrientation: Validator = (value) => value === 0 || value === 1;

// These are the actual action creator functions

const updatePositionX = kitActionCreator(isValidPositionX)('positionX');

const updatePositionY = kitActionCreator(isValidPositionY)('positionY');

const updateScale = kitActionCreator(isValidScale)('scale');

const updateRotation = kitActionCreator(isValidRotation)('rotation');

const updateLayer = kitActionCreator(isValidLayer)('layer');

const updateOrientation = kitActionCreator(isValidOrientation)('orientation');

export {
  updatePositionX,
  updatePositionY,
  updateScale,
  updateRotation,
  updateLayer,
  updateOrientation,
};
