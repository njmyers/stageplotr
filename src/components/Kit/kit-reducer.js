import type { Action } from './kit-actions';
import type { Action as MapAction } from '../Map/map-actions';

export type Kit = {
  /** a number id specifying which piece of kit (mapped to graphics/metadata) */
  typeID: number,
  /** number specifying the category of the kit */
  category: number,
  /** whole number pixel value for position on y axis */
  positionX: number,
  /** whole number pixel value for position on x axis */
  positionY: number,
  /** scale is expressed as a whole number percentage */
  scale: number,
  /** orientation is a boolean expressing whether or not the image is flipped */
  orientation: boolean,
  /** rotation is a whole number in degrees ie max 360 */
  rotation: number,
  /** layer is a whole number expressing z-index style properties */
  layer: number,
};

export const initialKit: Kit = {
  typeID: 0,
  category: 0,
  positionX: 0,
  positionY: 0,
  scale: 100,
  orientation: false,
  rotation: 0,
  layer: 1,
};

function kitReducer(state: Kit = initialKit, action: Action & MapAction) {
  switch (action.type) {
    case '@KIT/UPDATE_PROPERTY':
      return {
        ...state,
        [action.key]: action.value,
      };
    default:
    case '@MAP/CREATE_KIT':
      return {
        ...initialKit,
        typeID: action.typeID,
      };
      return state;
  }
}

export default kitReducer;
