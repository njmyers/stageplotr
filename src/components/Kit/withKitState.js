// @flow
import * as React from 'react';
import { connect } from 'react-redux';
// actions
import {
  updatePositionX,
  updatePositionY,
  updateScale,
  updateRotation,
  updateLayer,
  updateOrientation,
} from './kit-actions';

type Props = {
  id: string,
};

/**
 * HOC to wrap a component with the appropriate kit values/methods
 * Since kits are stored in a map as a hash table you must also pass in id.
 * This will access the correct hashed object from the table and subscribe
 * your component to the redux updates.
 */
const withKitState = (id: string) => (Wrapped: React.ComponentType<Props>) => {
  const mapStateToProps = (state) => ({
    ...state.map.kits[id],
  });

  const mapDispatchToProps = (dispatch) => ({
    updatePositionX: (value) => dispatch(updatePositionX(value, id)),
    updatePositionY: (value) => dispatch(updatePositionY(value, id)),
    updateScale: (value) => dispatch(updateScale(value, id)),
    updateRotation: (value) => dispatch(updateRotation(value, id)),
    updateLayer: (value) => dispatch(updateLayer(value, id)),
    updateOrientation: (value) => dispatch(updateOrientation(value, id)),
  });

  return connect(
    mapStateToProps,
    mapDispatchToProps
  )(Wrapped);
};

export default withKitState;
