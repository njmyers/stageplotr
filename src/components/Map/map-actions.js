export const newKit = () => ({
  type: '@MAP/CREATE_KIT',
});

export const deleteKit = (value: string) => ({
  type: '@MAP/DELETE_KIT',
});
