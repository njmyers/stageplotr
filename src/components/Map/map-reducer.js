// @flow
import kitReducer from '../Kit/kit-reducer';
import type { Kit } from '../Kit/kit-reducer';
import type { Action } from './map-actions';

export type Map = {
  name: string,
  width: number,
  height: number,
  length: number,
  /** Jeff you were right the collection of kits needs to be something
   * besides a simple array
   * This currently is an implementation kind of like a map hash table.
   * This is flow syntax to designate that behavior
   * unknown property key but specific object type
   * https://flow.org/en/docs/types/objects/#toc-objects-as-maps
   */
  kits: {
    [key: string]: Kit,
    __proto__: null,
  },
};

const initialState = {
  name: 'StagePlot',
  width: 100,
  height: 100,
  length: 0,
  // https://davidwalsh.name/object-create-null
  // blank dictionary
  kits: Object.create(null),
};

function mapReducer(state: Map = initialState, action: Action) {
  switch (action.type) {
    // these are integrated with kits
    case '@MAP/CREATE_KIT':
      return {
        ...state,
        length: state.length + 1,
        kits: {
          ...state.kits,
          [action.id]: kitReducer({}, action),
        },
      };
    case '@MAP/DELETE_KIT': {
      // functional object deleting
      const { [action.id]: deleted, ...rest } = state.kits;
      return {
        ...state,
        length: state.length - 1,
        kits: {
          ...rest,
        },
      };
    }
    case '@KIT/UPDATE_PROPERTY': {
      const { [action.id]: updateable, ...rest } = state.kits;
      return {
        ...state,
        kits: {
          ...rest,
          [action.id]: kitReducer(updateable, action),
        },
      };
    }
    // these are straight map
    case '@MAP/RESIZE_WIDTH':
      return {
        ...state,
        width: action.value,
      };
    case '@MAP/RESIZE_HEIGHT':
      return {
        ...state,
        height: action.value,
      };
    case '@MAP/RENAME':
      return {
        ...state,
        name: action.value,
      };

    default:
      return state;
  }
}

export default mapReducer;
