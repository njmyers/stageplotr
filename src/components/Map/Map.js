// @flow
import * as React from 'react';
import uuid from 'uuid/v1';

import Kit from '../Kit';
import Resize from '../Resize';
import type { State as MapState } from './map-reducer';

type State = {};

type Props = {
  map: MapState,
};

class Map extends React.Component<Props, State> {
  svg = null;

  onCreate = () => {
    this.props.createKit();
  };

  render() {
    return (
      <section>
        <svg
          ref={(svg) => (this.svg = svg)}
          width={this.props.map.width}
          height={this.props.map.height}
        >
          {Object.keys(this.props.map.kits).map((id) => {
            return (
              <Resize key={id}>
                <Kit id={id} />
              </Resize>
            );
          })}
        </svg>
        <button onClick={this.onCreate}>Create Kit!</button>
      </section>
    );
  }
}

export default Map;
