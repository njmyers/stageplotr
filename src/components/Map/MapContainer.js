import * as React from 'react';
import { connect } from 'react-redux';
import Map from './Map';

const mapStateToProps = (state) => ({});

const mapDispatchToProps = (dispatch) => ({});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Map);
