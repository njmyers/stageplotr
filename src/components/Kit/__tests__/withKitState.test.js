import * as React from 'react';
import { shallow } from 'enzyme';
import withShallowStore from '../../__tests__/helpers/withShallowStore';

import { initialKit } from '../kit-reducer';
import withKitState from '../withKitState';
import { createKit } from '../../Map/map-actions';
import { updateScale } from '../kit-actions';

import { createStore } from 'redux';
import rootReducer from '../../Plotr/root-reducer';

const id = '928d2ca0-94e4-11e8-82e4-61f5917152d6';

const store = createStore(rootReducer);
store.dispatch(createKit(id, 0));

const Kit = (props) => null;
const ConnectedKit = withKitState(id)(Kit);

describe('integration test withKitState HOC', () => {
  test('sanity test it renders without crashing', () => {
    const wrapper = withShallowStore(<ConnectedKit />, store);
    expect(wrapper.props()).toMatchSnapshot();
  });

  test('it renders initial kit props', () => {
    const wrapper = withShallowStore(<ConnectedKit />, store);
    expect(wrapper.props()).toMatchObject({ ...initialKit });
  });

  test('the component receives props from dispatched actions', () => {
    store.dispatch(updateScale(110, id));
    const wrapper = withShallowStore(<ConnectedKit />, store);
    expect(wrapper.props().scale).toBe(110);
  });

  test('the component creates actions from react methods', () => {
    const wrapper = withShallowStore(<ConnectedKit />, store);
    expect(wrapper.props().updateScale(120)).toMatchObject({
      id,
      key: 'scale',
      type: '@KIT/UPDATE_PROPERTY',
      value: 120,
    });
  });
});
