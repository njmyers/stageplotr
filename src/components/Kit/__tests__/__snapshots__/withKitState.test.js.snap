// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration test withKitState HOC sanity test it renders without crashing 1`] = `
Object {
  "category": 0,
  "layer": 1,
  "orientation": 0,
  "positionX": 0,
  "positionY": 0,
  "rotation": 0,
  "scale": 100,
  "typeID": 0,
  "updateLayer": [Function],
  "updateOrientation": [Function],
  "updatePositionX": [Function],
  "updatePositionY": [Function],
  "updateRotation": [Function],
  "updateScale": [Function],
}
`;
